### ISSUE TITLE
foreach loop variables should not be assignable

### ISSUE THREAD

**Issue by bob-carpenter** [stan-dev/stan#2527]
#### Summary:

Apparently elements of a for-each loop can be assigned.  Instead, it should be have the same way as the basic for loops and not allow the looped variable to be assigned.  

#### Description:

The general pattern in Stan is to disallow reassignment of arguments like these (e.g., function arguments).

I've already expunged mention of assignability from the yet-to-be-released 2.18 doc, but we need to change the code to throw an error if the loop variable is assigned.  The regular for loop should indicate how to do this.

#### Current Version:
develop (pre-2.18)

**Comment by bob-carpenter** [comment]
@VMatthijs  I understand the example you had, 

```
for (y in ys) y = z;
```

which is intended to set each element of `ys` equal to `z`, but we don't have this kind of reference-like behavior anywhere else in Stan, so I don't think we should allow it.

**Comment by mitzimorris** [comment]
given current implementation, adding additional type to `stan/lang/ast/origin_block` specific to loop variables and updating logic in `stan/lang/ast/scope` should make this work cleanly.

**Comment by mitzimorris** [comment]
made changes, now parse fails w/ error msg:
```
C++ exception with description "SYNTAX ERROR, MESSAGE(S) FROM PARSER:

Cannot assign to variable outside of declaration block; left-hand-side variable origin=loop identifier
  error in 'utility-stub.stan' at line 5, column 22
  -------------------------------------------------
     3:     int vs[2];
     4:     int y;
     5:     for (v in vs) v = 3;
                             ^
     6:     return 0;
  -------------------------------------------------

PARSER EXPECTED: <expression assignable to left-hand side>
```
@bob-carpenter , @VMatthijs - opinions?

**Comment by bob-carpenter** [comment]
It would be nice if the caret pointed at the variable that shouldn't be assigned, but the real problem is that it shouldn't have that PARSER EXPECTED bit.  Togehter, they make it look like the problem is the right-hand side, not the variable on the left.

**Comment by mitzimorris** [comment]
> it shouldn't have that PARSER EXPECTED bit

we could modify parser error messages so that it only prints the PARSER EXPECTED bit when there aren't any other error messages generated by the parser.  that is, for input semantically well-formed up to a point,  the "PARSER EXPECTED" message is (sometimes) useful.

**Comment by mitzimorris** [comment]
> It would be nice if the caret pointed at the variable that shouldn't be assigned
this should be its own issue.  the placement of the caret determined by semantic action function `program_error_f`.  it knows how far the parser has read into the program; doesn't know how far to backtract to get to relevant part of expression which has an error.

**Comment by VMatthijs** [comment]
Hi @bob-carpenter, thanks for spotting that! That is indeed inconsistent with the rest of the language. Not sure why I thought that would be a good idea. I guess it gives you a little bit of extra expressive freedom and it was easy to set up, so I got tempted. I agree though that consistency is more important.

If it is not fixed before I get to New York, I'm happy to fix it then. :-)

**Comment by mitzimorris** [comment]
@VMatthijs  got an almost working branch here.

**Comment by mitzimorris** [comment]
@VMatthijs would you be able to review PR https://github.com/stan-dev/stan/pull/2535 for me?

**Comment by mitzimorris** [comment]
fixed by PR #2535 
closing.
